VSR: SuperBlock VSR recovery

sentientwaffle opened the pull request:
### Summary

Replace recovery protocol with the superblock.

### Details

Currently when a replica recovers from a crash, it switches to `status=recovering` and starts "recovery protocol" (as described in [VRR](https://pmg.csail.mit.edu/papers/vr-revisited.pdf) §4.3). This allows the replica to safely rejoin the cluster's current (active) view. It has a major drawback though: for recovery protocol to complete, there must be a full view-change quorum (4/6 etc) of replicas in `status=normal`. Which means that if all replicas are restarted simultaneously, the cluster can never recover.

Instead of using recovery protocol:
- Before sending a message in a particular view, ensure that that view is persisted to the superblock.
- Before sending a DVC/SV, ensure that the DVC's/SV's headers are persisted to the superblock.

In combination, these guarantee that if a replica isolates an old primary, it remains isolated after the replica restarts. That is, it is impossible for a replica to participate in view _A_, then view _B_, then view _A_ again.

The "recovery" logic lives at the end of `Replica.open()`:

    if (self.replica_count == 1) {
        if (self.journal.faulty.count > 0) {
            @panic("journal is corrupt");
        }
        assert(self.op_head_certain());

        if (self.commit_min < self.op) {
            self.commit_journal(self.op);
        } else {
            self.transition_to_normal_from_recovering_status();
        }
    } else {
        if (self.op_head_certain()) {
            if (self.view == self.log_view) {
                if (self.primary_index(self.view) == self.replica) {
                    self.transition_to_view_change_status(self.view + 1);
                } else {
                    self.transition_to_normal_from_recovering_status();
                }
            } else {
                assert(self.view > self.log_view);
                self.transition_to_view_change_status(self.view);
            }
        } else {
            self.transition_to_recovering_head();
        }
    }


`recovering_head` is a new replica status for when the head op is "uncertain" (see `Replica.op_head_certain()`). In this status, the replica keeps requesting a `start_view` message, which will resolve its uncertainty.


## Pre-merge checklist

Performance:

* [ ] Compare `zig benchmark` on linux before and after this pr.
    ``` sh
    # benchmark results before
    ...
    
    # benchmark results after
    ...
    ```
OR
* [x] I am very sure this PR could not affect performance.
    * (This impacts view-change performance, but does not impact the prepare/commit path.)

sentientwaffle committed:
VSR,SuperBlock: {view,log_view}_durable

Before sending any message with view=X, view=X must be durable (within the superblock sector).
Advancing the view isolates the primary, and persisting the advanced view ensures that this isolation is maintained after a replica restarts.

The implementation approach is simple and a bit brute-force: in `send_message_to_replica()`, drop outgoing messages with a `view` (or in the case of a DVC, the `log_view`) which exceeds the durable `view` (or `log_view`).
To keep things simple, after the `view`/`log_view` is bumped, the replica doesn't try to immediately resume whatever was blocked by the superblock update.
Dropped messages will be eventually retried, either because the sending replica has a timeout (e.g. DVC), or because another replica will drive a retry (e.g. `prepare` → `prepare_ok`).

sentientwaffle committed:
VSR: Rework create_view_change_message

(Preparing to move the header generation logic into `view_durable_update()`.)

sentientwaffle committed:
VSR: Store view-change headers in superblock

The "SuperBlockSector" is no longer a single sector — it is at least 2, and may be more depending on the pipeline prepare queue size (but usually 2).

sentientwaffle committed:
VSR: Remove recovery protocol

Use `SuperBlockSector.vsr_headers()` instead of recovery protocol to recover.

Two known bugs:

1. In `repair()`, `journal.find_latest_headers_break_between()` expects `commit_min` to exist, but when `commit_min == op_checkpoint` this is not necessarily the case.
   (I already knew about this bug but can now — at long last — reproduce in the simulator):
2. We need to use `op_certain()` to avoid losing the head op.

sentientwaffle committed:
VSR: Repair to commit_min+1

In replica repair we call `journal.find_latest_headers_break_between(replica.commit_min, replica.op)` — the journal here expects that both `commit_min` and `op` exist in the log. `replica.op` does exist, but `commit_min` might not. (This happens in exactly one case: `commit_min == op_checkpoint` and the checkpoint was overwritten; i.e. the entire log was filled without any messages being committed).

There are a couple possible fixes:

1. Add the header for `op_checkpoint + 1` to the superblock's VSR state; load it into the journal at replica open. Then we use `find_latest_headers_break_between(replica.commit_min + 1`, which would always exist.
2. Change `find_latest_headers_break_between` to allow its `op_min` argument to be missing (which is reported as a break), and again use `find_latest_headers_break_between(replica.commit_min + 1`.

I went with the latter since it feels like less of a kludge. The "op_min must exist" invariant in that function isn't actually useful if the op_min doesn't always exist.

sentientwaffle committed:
VSR: Add recovering_head status

When a replica starts up and there is a corrupt WAL entry "to the right" of its maximum op, it does not know the head for certain (`!op_head_certain()`).
In this case we must await a `start_view` message to avoid forking the log.

sentientwaffle committed:
VSR: Fix off-by-one in assertion

op_head - commit_min ≤ slot_count

Example:

Suppose slot_count=6 and batch_multiple=2.
Then the first wrap of the log is:

    0 1 2 3 4 5

And

    op_checkpoint_trigger = 5

After the checkpoint:

    op_checkpoint = 5 - batch_multiple = 3

The next wrap:

    6 7 8 9

And

    op_checkpoint_trigger=9

Then

    op_checkpoint_trigger - op_checkpoint = 9 - 3 = 6 ≤ slot_count

sentientwaffle committed:
VSR: More superblock validation

sentientwaffle committed:
Test: Update fuzzers with Superblock API change

sentientwaffle committed:
VOPR: Update crash guards

`Cluster.crash_replica()` now always crashes the replica. Since recovery no longer requires recovery protocol, any replica can be crashed at any time.

The only thing to be careful of is making sure that enough replicas have `op_head_certain()`, so that they don't get stuck in `recovering_head` mode. The simulator handles this, so that tests built on Cluster can test liveness issues.

Increase probability of crashes 20x. (May need to tune this some. Crashing *too* often will lead to the simulation running out of ticks).

sentientwaffle committed:
Test/Storage: Reduce superblock faults

Avoid a case where:

    B₀ (ok),      manifest (ok)
    B₁ (corrupt), manifest (ok)
    A₂ (ok),      manifest (corrupt)
    A₃ (ok),      manifest (torn)

The alternative is to repair trailers more aggressively, but that would require reading them all on `open()`, instead of just the first valid one.

sentientwaffle committed:
VSR: Repair corrupt root prepare

sentientwaffle committed:
VSR: Correct head after recovery

Guard against truncated ops "leaking" back into the journal after a crash.
For example:

1. Prepare ops 10, 11, 12, 13
2. View change. Truncate ops 11, 12, 13.
3. Crash. Recover.
4. At recovery, our head must be op 10, not op 13 — even though op 13 is higher and has not yet been overwritten.

To accomplish this, use `ViewChangeHeaders.view_for_op()` combined with the `log_view` (from the superblock's VSR state) to filter out ops in the journal which cannot possibly be part of the current `log_view`.
Only after these leaked logs are discarded is the `op_head` chosen.

This filtering must take place within the journal, as part of recovery, to ensure that the recovery case uses the correct `op_max`.

Lastly, the old case `@G` has been removed (merged into `@H`, which is now `@G` since they all shifted back). `@G`'s old explanation:

    /// One of:
    ///
    /// * A misdirected read to a reserved header.
    /// * The redundant header's write was lost or misdirected.
    ///
    /// For multi-replica clusters, don't repair locally to prevent data loss in case of 2 lost writes.

We actually do need to repair locally for `R>1`: We can end up in this case after a view change because truncated headers are replaced in-memory by `reserved`, which can then be written to disk as part of a redundant-header sector. And we need to avoid `decision=vsr` in this case since it can lead to `!op_head_certain` → `status=recovering_head`, which deadlocks a `R=2` cluster.
(Double-lost-writes aren't really in our fault model anyway, that was just a nice-to-have).

sentientwaffle committed:
VSR: Correct DVC after recovery

It is important that the headers included in a DVC are correct, since otherwise the view-change can get messed up. We store those headers in the superblock before ever sending a DVC. (This is what lets us not need recovery protocol).
When we recover after a crash, we may (depending on our superblock's `view`/`log_view`) need to start retrying a DVC that we persisted just before crashing. The question is: how should we use the superblock headers to accomplish this?

There are two options:

1. At startup our Journal may have errors, e.g. corruption or incorrect entries. We could replace entries in the journal with headers from the DVC. We would also need to remove entries from the journal to mirror gaps in our DVC. If this is done carefully, we should be able to recreate the exact DVC using our journal. (At least I think so — this is super hard to reason about near checkpoint boundaries, where safe header replacement gets tricky).
2. Or, if we need to resend the DVC, we could just see that we already generated it (since it is part of the working superblock (`vsr_headers`)) and send those exact headers, rather than constructing the DVC from our journal.

I had a hard time choosing between these approaches, but ended up going with option 2 because even though it feels a little like a hack (b/c it is less symmetrical with the start-view case), it seems much harder to get wrong — it is "obviously correct". The "replace headers" and "mirror gaps" steps from option 1 are both subtle.

(We still do need to do the "replace" step b/c our superblock headers might be SV headers, not DVC headers, but the SV is much more forgiving — all we actually need in that case (for correctness) is the head op. Any extra headers beyond that are just bonus to reduce the need for repair).

sentientwaffle committed:
VOPR: Work around journal zeroed redundant headers

Crash injection requires an additional guard to prevent the cluster getting stuck in `status=recovering_head`.

sentientwaffle committed:
VSR: Fix view_durable assertion failure

If a replica recovers into `status=recovering_head`, then receives a `start_view` for the same view it was in before the crash, we don't need to update `view_durable`/`log_view_durable` in the superblock.

sentientwaffle committed:
Fuzz: Fix superblock fuzzer

sentientwaffle committed:
Test/Storage: Fix off-by-one

sentientwaffle committed:
VSR: zeroInit() → zeroes()

sentientwaffle committed:
VSR: Verify DVC/SV header checksums

sentientwaffle commented on src/vsr/journal.zig:
```diff
@@ -1179,17 +1144,17 @@ pub fn JournalType(comptime Replica: type, comptime Storage: type) type {
+        ///   match op                 _   _   _   _   _   _   _   _  !1   <   >   1  !1
+        ///   match view               _   _   _   _   _   _   _   _  !1   _   _  !0  !1
+        ///   decision (replicas>1)  vsr vsr vsr vsr vsr fix fix vsr nil fix vsr vsr eql
+        ///   decision (replicas=1)              fix fix
```
The diff here is confusing; I just removed the `@G` column, and then re-labeled everything after it (`@H` → `@G`, `@I` → `@H`, etc).

sentientwaffle commented on src/vsr/journal.zig:
```diff
@@ -1235,7 +1233,7 @@ pub fn JournalType(comptime Replica: type, comptime Storage: type) type {
 
                 // `prepare_checksums` improves the availability of `request_prepare` by being more
                 // flexible than `headers` regarding the prepares it references. It may hold a
-                // prepare whose redundant header is broken, as long as the prepare itjournal is valid.
```
A while back I did `%s/self/journal` and missed these typos :facepalm:

sentientwaffle commented on src/vsr/journal.zig:
```diff
@@ -1451,11 +1450,15 @@ pub fn JournalType(comptime Replica: type, comptime Storage: type) type {
-                    journal.headers[dirty_slot].checksum);
-                assert(journal.prepare_checksums[dirty_slot] ==
-                    journal.headers_redundant[dirty_slot].checksum);
+                if (journal.prepare_inhabited[dirty_slot]) {
```
This was a preexisting bug, discovered by the VOPR when the crash injection was turned up.

sentientwaffle commented on src/vsr/superblock.zig:
```diff
@@ -32,6 +32,9 @@ pub const Quorums = @import("superblock_quorums.zig").QuorumsType(.{
+    ((constants.view_change_headers_max * @sizeOf(vsr.Header)) % constants.sector_size);
+
 // Fields are aligned to work as an extern or packed struct.
 pub const SuperBlockSector = extern struct {
```
I'm not sure whether we should rename `SuperBlockSector` now that it is (at minimum) 2 sectors in size. Maybe `SuperBlockHeader`?

jamii replied:
Sounds reasonable.

sentientwaffle replied:
:+1: I'll rename in a separate PR.

sentientwaffle commented on src/vsr/superblock.zig:
```diff
@@ -258,17 +276,17 @@ pub const SuperBlockSector = extern struct {
         if (a.manifest_size != b.manifest_size) return false;
         if (a.free_set_size != b.free_set_size) return false;
-
-        assert(@bitCast(u32, a.reserved[0..4].*) == 0);
-        assert(@bitCast(u32, b.reserved[0..4].*) == 0);
```
These aren't needed anymore since `reserved`'s size is now a multiple of 8-bytes.

sentientwaffle commented on src/vsr/superblock.zig:
```diff
@@ -662,21 +688,33 @@ pub fn SuperBlockType(comptime Storage: type) type {
+        const UpdateCheckpoint = struct {
+            commit_min_checksum: u128,
+            commit_min: u64,
+            commit_max: u64,
+        };
+
         /// The vsr_state must update the commit_min and commit_min_checksum.
-        // TODO Will the replica ever update view/log_view by calling checkpoint() during a view
-        // change? If not, forbid it.
+        /// The vsr_state must not update view/log_view.
         pub fn checkpoint(
             superblock: *SuperBlock,
             callback: fn (context: *Context) void,
             context: *Context,
-            vsr_state: SuperBlockSector.VSRState,
+            update: UpdateCheckpoint,
         ) void {
```
Previously `SuperBlock.view_change()` and `SuperBlock.checkpoint()` took a `VSRState` parameter, and then they did a bunch of assertions to make sure that only the correct fields were updated. (`checkpoint` updates commits, `view_change` updates view numbers). Now only the parameters that _can_ be modified are passed in.

sentientwaffle commented on src/vsr/superblock.zig:
```diff
@@ -1236,12 +1309,14 @@ pub fn SuperBlockType(comptime Storage: type) type {
                 // We do not repair padding.
                 context.copy = 0;
                 superblock.read_free_set(context);
-            } else if (copy + 1 == constants.superblock_copies) {
-                @panic("superblock manifest lost");
             } else {
                 log.debug("open: read_manifest: corrupt copy={}", .{copy});
-                context.copy = copy + 1;
-                superblock.read_manifest(context);
+                if (copy + 1 == constants.superblock_copies) {
+                    @panic("superblock manifest lost");
+                } else {
+                    context.copy = copy + 1;
+                    superblock.read_manifest(context);
+                }
             }
```
This wasn't a bug; it is just nice to get that same log message for each corrupt copy, instead of all except the last.

sentientwaffle commented on src/testing/cluster.zig:
```diff
@@ -301,99 +301,17 @@ pub fn ClusterType(comptime StateMachineType: fn (comptime Storage: type, compti
         /// Returns whether the replica was crashed.
         /// Returns an error when the replica was unable to recover (open).
-        pub fn crash_replica(cluster: *Self, replica_index: u8) !bool {
+        pub fn crash_replica(cluster: *Self, replica_index: u8) !void {
```
`crash_replica` used to only crash the replica if the cluster could ever recover.
It is now the caller's responsibility to check whether the cluster can ever recover before calling the function. (This is _much_ simpler than before, though).

sentientwaffle commented on src/vsr/replica.zig:
```diff
@@ -3026,17 +2692,26 @@ pub fn ReplicaType(
         ///
         /// The caller owns the returned message, if any, which has exactly 1 reference.
         fn create_view_change_message(self: *Self, command: Command) *Message {
```
We create the message, but then only send it if the `view` matches the `view_durable` (and in the case of DVC, if the `log_view` matches the `log_view_durable`. (See `send_message_to_replica()`).

There is a bit of subtlety here, though: We synchronously send the message to _ourself_ (when we are the primary), even if the view/log_view is not durable yet. This simplifies the `on_do_view_change()`, which expects to always see a DVC from itself.

sentientwaffle commented on src/vsr/replica.zig:
```diff
@@ -3051,65 +2726,120 @@ pub fn ReplicaType(
-                if (command == .start_view) {
-                    // The primary starting a new view has a pristine log suffix — no gaps or breaks.
-                    while (op > self.commit_min) : (op -= 1) {
+        fn create_view_change_headers(self: *const Self) vsr.ViewChangeHeaders.BoundedArray {
```
This is essentially the same logic from before (in `create_view_change_message()`) with a few changes:

- The `Ensure that if we started a DVC before a crash` block is new.
- In the start-view section, require 1 fewer header, since `commit_min` does not necessarily exist in the journal. (When `commit_min == op_checkpoint`, and `op_checkpoint` is overwritten or corrupt).
- Split the start-view section since it is called by in more situations (e.g. followers updating their superblock headers). In these cases we just include has much chained suffix as possible — all we strictly need is the head op, so this is quite forgiving.
- Fix Retired backup DVC generation (it previously skipped headers that should be been included, but in a way that didn't violate correctness, so it accidentally worked anyway :facepalm:).

sentientwaffle commented on src/vsr/journal.zig:
```diff
@@ -613,7 +583,6 @@ pub fn JournalType(comptime Replica: type, comptime Storage: type) type {
             op_max: u64,
         ) ?HeaderRange {
             assert(journal.status == .recovered);
-            assert(journal.header_with_op(op_min) != null);
```
We used to require that `op_min` exists, but the replica can't make that guarantee.

sentientwaffle commented on src/vsr/replica.zig:
```diff
@@ -3970,36 +3659,43 @@ pub fn ReplicaType(
-                    self.commit_min,
-                    self.commit_max,
+            if (self.commit_min != self.op) {
+                var broken = self.journal.find_latest_headers_break_between(
```
`commit_min` usually exists, but _not always_. And if it doesn't exist, we don't actually want to repair it: the case were it doesn't exist is when it is `op_checkpoint` and has been overwritten (or may have been overwritten — if there is corruption we might not be able to tell which).

sentientwaffle commented on src/vsr/replica.zig:
```diff
@@ -4393,7 +4089,7 @@ pub fn ReplicaType(
             assert(self.repairs_allowed());
             assert(self.journal.dirty.count > 0);
             assert(self.op >= self.commit_min);
-            assert(self.op - self.commit_min + 1 <= constants.journal_slot_count);
```
An off-by-one error that the VOPR found.

sentientwaffle commented on src/vsr/replica.zig:
```diff
@@ -5097,6 +4771,42 @@ pub fn ReplicaType(
+                // Critical: Do not advertise a view/log_view before it is durable.
+                // See view_durable()/log_view_durable().
+                if (message.header.view > self.view_durable() and
+                    message.header.command != .request_start_view)
```
`request_start_view` includes the view being requested. We don't need (or want) to persist that view until we have received the start-view in question.

jamii commented:
https://github.com/tigerbeetledb/tigerbeetle/blob/800185e2b4f38b8fdd59e4a0f1d6799996adde01//src/simulator.zig#L262

Why do we not test repeated crashes?

sentientwaffle commented:
> Why do we not test repeated crashes?

Since the stability is set randomly:

```
        .replica_crash_probability = 0.00002,
        .replica_crash_stability = random.uintLessThan(u32, 1_000),
        .replica_restart_probability = 0.0001,
        .replica_restart_stability = random.uintLessThan(u32, 1_000),
```

I think that it does test it ok.

jamii commented on src/vsr.zig:
```diff
@@ -1026,3 +998,88 @@ pub fn quorums(replica_count: u8) struct {
+    };
+
+    /// Returns the range of possible views (of prepare, not commit) for a message that is part of
+    /// the same log_view as these headers.
```
I'm not totally clear on what this function is doing eg I can't parse 'part of the same log_view' - does that just mean 'if the log didn't fork'?.

sentientwaffle replied:
So in the caller context, we have some headers (the superblock's VSR headers) and a `log_view` (the superblock's `log_view`), and those correspond — the VSR headers are either SV or DVC headers corresponding to that `log_view`.

The caller (in `journal.zig`) wants to check for each op in the WAL, whether that op could _possibly_ belong to the superblock's `log_view`. (If it cannot, the op is discarded).

For example, suppose we have this start-view message in our superblock, along with `log_view=c`:
(Numbers indicate ops, letters indicate the view during which the op was prepared)

    3b 4b 5b 6b

- If we have a message `2b` in our WAL, it might be part of `log_view=c`.
- If we have a message `2a` in our WAL, it might be part of `log_view=c`. (`a < b`).
- If we have a message `2c` in our WAL, it cannot possibly be part of `log_view=c`, since `c > b` — `3b` _is_ part of the log, and was prepared before `2c`.
- If we have a message `7b` in our WAL, it cannot possibly be part of `log_view=c`, since if it was, it would have been included in the SV message to begin with.

jamii commented on src/vsr.zig:
```diff
@@ -1026,3 +998,88 @@ pub fn quorums(replica_count: u8) struct {
+        const header_oldest = &headers.slice[headers.slice.len - 1];
+
+        if (op < header_oldest.op) return .{ .min = 0, .max = header_oldest.view };
+        if (op > header_newest.op) return .{ .min = log_view, .max = log_view };
```
I'm suprised this isn't `.min = header_newest.view`. Why are we sure that ops after the headers must have been from the current view? Because we would have saved them to the superblock if they were from before log_view?

sentientwaffle replied:
> Why are we sure that ops after the headers must have been from the current view?

The actual headers in question here are either:
- DVC headers, or
- SV headers
for a particular `log_view`.

To answer your question for each separately:
- DVC: If we have the DVC headers for a `log_view` `V`, we must be in `view_change` status trying to transition to some view greater than `V`. So we will never prepare anything else as part of view `V`. So `header_newest.view` would actually work too (in the DVC case).
- SV: If we have the SV headers for a `log_view`, we can continue add to them (preparing more ops), but those ops will always be part of the `log_view`. ("Because we would have saved them to the superblock if they were from before log_view?" is exactly right).

jamii replied:
Could you add this explanation and the above one as comments?

sentientwaffle replied:
Rewritten to:
```
    /// Returns the range of possible views (of prepare, not commit) for a message that is part of
    /// the same log_view as these headers.
    ///
    /// - When these are DVC headers for a log_view=V, we must be in view_change status working to
    ///   transition to a view beyond V. So we will never prepare anything else as part of view V.
    /// - When these are SV headers for a log_view=V, we can continue to add to them (by preparing
    ///   more ops), but those ops will laways be part of the log_view. If they were prepared during
    ///   a view prior to the log_view, they would already be part of the headers.
```

jamii commented on src/vsr.zig:
```diff
@@ -1026,3 +998,88 @@ pub fn quorums(replica_count: u8) struct {
+            view_max = header.view;
+        } else {
+            unreachable;
+        }
```
Kinda picky, but it might be easier to read without `view_max`.

```suggestion
        for (headers.slice) |*header| {
            if (header.op == op) return .{ .min = header.view, .max = header.view };
        }
        for (headers.slice[0..headers.slice.len-1]) |*header_prev, header_prev_index| {
            const header_next = headers.slice[header_prev_index+1];
            if (header_prev.op < op and op < header_next.op) return .{.min = header_prev.view, .max = header_next.view};
        }
        unreachable;
```

sentientwaffle replied:
Thanks! Minor change to swap next/prev since headers are ordered from high-to-low op.

jamii replied:
Ha, I did misread the original version :D

jamii commented on src/vsr/journal.zig:
```diff
@@ -1212,13 +1177,46 @@ pub fn JournalType(comptime Replica: type, comptime Storage: type) type {
+                }
+            }
+
+            for (journal.headers) |*header_untrusted, index| {
```
Can we combine these loops?

```
for (.{journal.headers_redundant, journal.headers}) |headers| {
   ...
}
```

sentientwaffle replied:
Yes, though it ends up being
```
            for ([_][]align(constants.sector_size) Header{
                journal.headers_redundant,
                journal.headers,
            }) |headers| {
```
to satisfy a zig type inference bug.

jamii commented on src/vsr/superblock_fuzz.zig:
```diff
@@ -310,14 +315,29 @@ const Environment = struct {
             .view = env.superblock.staging.vsr_state.view + 5,
         };
 
+        var vsr_headers = vsr.ViewChangeHeaders.BoundedArray{ .buffer = undefined };
```
```suggestion
        var vsr_headers = vsr.ViewChangeHeaders.BoundedArray.init(0);
```

sentientwaffle replied:
I avoid `BoundedArray.init()` because it returns an error based on the length passed in, so it would need to be `catch unreachable`.

sentientwaffle commented:
The simulator found a view-change bug; I'm tracking it down now.

jamii commented on src/vsr/replica.zig:
```diff
@@ -370,33 +350,62 @@ pub fn ReplicaType(
+            }
+
+            self.op = op_head;
+            for (vsr_headers.slice) |*header| self.replace_header(header);
```
Why do we need to replace on open? Do we not ensure that the journal matches before writing the superblock?

sentientwaffle replied:
> Do we not ensure that the journal matches before writing the superblock?

No — the WAL can lag behind the in-memory journal, but the in-memory journal is lost during a crash.

The in-memory journal is recovered from the WAL at startup, but it isn't necessarily accurate. In particular, the journal is only updated when we write prepares, but the journal is updated with headers from view changes. So we may receive some headers in a `start_view`, but not have those in our WAL. Or there may be entries in our WAL which were truncated from the journal, which "leak through" during recovery — we try to remove these when we can identify them.

Of all of the headers replaced here (in this loop), only the first one is critical, since it might be our new head op, which _must_ exist in the in-memory journal headers. ("head op exists in journal headers" is a replica invariant.)

sentientwaffle replied:
> Of all of the headers replaced here (in this loop), only the first one is critical

All of the headers after that which we replace are just bonus — replacing them here can reduce the number of round-trips required for repair.

jamii commented on src/vsr/replica.zig:
```diff
@@ -370,33 +350,62 @@ pub fn ReplicaType(
+                        if (self.primary_index(self.view) == self.replica) {
+                            self.transition_to_view_change_status(self.view + 1);
+                        } else {
+                            self.transition_to_normal_from_recovering_status();
```
I'm curious why there's no `if (self.commit_min < self.op)` in here.

sentientwaffle replied:
I'm not sure what you mean exactly; can you elaborate?

sentientwaffle replied:
Oh, I see, because `replica_count == 1` has to do that.

sentientwaffle replied:
A cluster-of-1 is effectively the primary, which must be caught up on commits before it can start preparing more ops (modulo the pipeline).

But in this case, the replica is a follower — it can commit asynchronously while also preparing/acking ops for the primary.

sentientwaffle committed:
VSR: Code review suggestions

sentientwaffle committed:
VSR: Guard against bad nack

`replica.open()` was using `replica.replace_header()` (and in turn, `journal.set_header_as_dirty()`) to load the view-change headers into the journal.
But `set_header_as_dirty()` was written with view-change and header-repair in mind, so it cleared `faulty`, when in fact the faulty entry might be the header from the view change, in which case a replica could (incorrectly) nack in response to a `request_prepare`.

This was discovered by the VOPR when a `commit_max ≤ op_head` assertion failed after a view change.

sentientwaffle commented:
> The simulator found a view-change bug; I'm tracking it down now.

This was a tricky one to root-cause! But it made me realize that the head op is _always_ certain (even if there are faults in the WAL) when the superblock's view-change headers are from a DVC — we only have to deal with the SV case.

Fixed in https://github.com/tigerbeetledb/tigerbeetle/pull/426/commits/f583bbb1c021ebda0c1cd49443e0a47bb19b39a1

sentientwaffle commented:
In a bit more detail:

When a replica knows that a Prepare should exist, but it doesn't have a copy (or its copy is corrupt), it sends a `request_prepare` message to other replicas, asking for their copy.

When the requesting replica is a primary, repairing to set up a new view, replicas may respond with a nack (`command=nack_prepare`) to indicate that:
1. The replica doesn't have that prepare, and
2. the replica is certain that it never did

If the primary gets enough nacks, then it can conclude that the message cannot possibly be committed, and it can truncate that from the log.

> the replica is certain that it never did

If a replica nacks a prepare that it _did_ have, then the primary could truncate a message that was committed. So a replica must err on the side of not nacking. If a replica has a corrupt entry in its WAL, it must not nack any op that would land in that entry (since after recovering, it doesn't know which op that actually was).

The bug above was: While loading the view-change headers, we were erroneously clearing the faulty bit during `Replica.open()`, leading to nacks of committed ops.

sentientwaffle committed:
VSR: Remove dead code

jamii commented:
bors d+

sentientwaffle commented:
Thanks for the review!

bors r+

sentientwaffle commented:
bors r-

Just found a bug :facepalm:

sentientwaffle committed:
VSR: Remove incorrect assertion

The comment:

    // At least one replica in the new quorum committed in the new replica.op's WAL wrap —
    // wrapping implies a checkpoint (which implies a commit).

is correct, but at this stage, the new primary has not updated its `commit_max`, so the assertion is not correct.
(I think this is left over from when DVCs included the `commit_max`.
A while back this was swapped out for the `commit_min`.
But this bug is *really* hard to hit.)

sentientwaffle committed:
VSR: Persist primary SV after repair

Previously the primary of a new view would persist its view/log_view/vsr_headers immediately after updating its `log_view`, i.e. upon receipt of the DVC quorum.
But those headers are not necessarily canonical for the view.
The new primary must complete repairs before persisting the view-change data.
During repair, nacks may result in the log truncating, which changes the op head.

Persisting the headers before this truncation takes place can cause the primary to send the wrong (i.e. supposed-to-be-truncated) headers in a DVC after it crashes/recovers.

One downside to this approach is that `primary_start_view_as_the_new_primary()` doesn't/can't send out the SV message anymore, since at that point the `view_durable_update()` is running.
Another downside is that `view_durable_update()`'s latency is no longer hidden by the new primary's repair/commit. I think it would be safe to still hide it behind the commit latency, but the implementation is more complicated so I didn't do that.

sentientwaffle committed:
VSR/DVC: Anchor DVC to durable SV

Ensure that enough headers are sent in a DVC after recovering with `view_durable == log_view_durable`.
I think there is a way to use this insight to simplify `create_view_change_headers()`, but saving that for a follow-up commit — this fixes the bug.

sentientwaffle commented:
bors r+
